nomos-da/network/core/src/maintenance/monitor.rs

```diff
@@ -0,0 +1,356 @@
+};
+use thiserror::Error;
+// internal
+
+#[derive(Copy, Clone, Debug, PartialEq)]
+pub enum PeerStatus {
+    Malicious,
+    Unhealthy,
+    Healthy,
+}
+
+pub struct ConnectionMonitorOutput {
+    pub peer_id: PeerId,
+    pub peer_status: PeerStatus,
+}
+
+pub trait ConnectionMonitor {
+    type Event;
+
+    fn record_event(&mut self, event: Self::Event) -> Option<ConnectionMonitorOutput>;
+    fn reset_peer(&mut self, peer_id: &PeerId);
+}
+
+/// A `NetworkBehaviour` that block connections to malicious peers or temporarily disconnects from
+/// unhealthy peers.
+#[derive(Debug)]
+pub struct ConnectionMonitorBehaviour<Monitor> {
+    monitor: Monitor,
+    malicous_peers: HashSet<PeerId>,
+    unhealthy_peers: HashMap<PeerId, Instant>,
+    close_connections: VecDeque<PeerId>,
+    redial_cooldown: Duration,
+    waker: Option<Waker>,
+}
+
+impl<Monitor> ConnectionMonitorBehaviour<Monitor>
+where
+    Monitor: ConnectionMonitor,
+{
+    pub fn new(monitor: Monitor, redial_cooldown: Duration) -> Self {
+        Self {
+            monitor,
+            malicous_peers: HashSet::new(),
+            unhealthy_peers: HashMap::new(),
+            close_connections: VecDeque::new(),
+            redial_cooldown,
+            waker: None,
+        }
+    }
+
+    fn try_wake(&mut self) {
+        if let Some(waker) = self.waker.take() {
+            waker.wake();
+        }
+    }
+
+    /// Block connections to a given peer.
+    ///
+    /// All active connections to this peer will be closed immediately.
+    ///
+    /// Returns whether the peer was newly inserted. Does nothing if the peer was already present in
+    /// the set.
+    pub fn block_peer(&mut self, peer: PeerId) -> bool {
+        let inserted = self.malicous_peers.insert(peer);
+        if inserted {
+            self.close_connections.push_back(peer);
+            self.try_wake();
+        }
+        inserted
+    }
+
+    /// **Temporarily blocks a peer** due to unhealthy behavior.
+    ///
+    /// The peer is added to `unhealthy_peers`, and after `redial_cooldown`, it will be allowed to reconnect.
+    /// The existing connections to this peer are **closed immediately**.
+    pub fn temporarily_block_peer(&mut self, peer: PeerId) {
+        let until = Instant::now() + self.redial_cooldown;
+        self.unhealthy_peers.insert(peer, until);
+        // Close existing connections
+        self.close_connections.push_back(peer);
+        self.try_wake();
+    }
+
+    /// Unblock connections to a given peer.
+    ///
+    /// Returns whether the peer was present in the set. Does nothing if the peer
+    /// was not present in the set.
+    pub fn unblock_peer(&mut self, peer: PeerId) -> bool {
+        let removed =
+            self.malicous_peers.remove(&peer) || self.unhealthy_peers.remove(&peer).is_some();
+        if removed {
+            self.try_wake();
+        }
+        removed
+    }
+
+    pub fn record_event(&mut self, event: Monitor::Event) {
```

Comment:
Wouldn't this be the only method that requires `Monitor` to implement `ConnectionMonitor`? Would it make sense to provide all the other methods for all the possible generics? I think it would make unit testing easier, but I do not have a strong opinion on this. I usually tend to do it, especially when the trait bound applies to a single method.

Reply:
Just to clarify, the suggested modification would look something like this?:
```
impl<Monitor> ConnectionMonitorBehaviour<Monitor> {
    pub fn new(monitor: Monitor, redial_cooldown: Duration) -> Self
    fn try_wake(&mut self)
    ....
}

impl<Monitor> ConnectionMonitorBehaviour<Monitor>
where
    Monitor: ConnectionMonitor,
{
    pub fn record_event(&mut self, event: Monitor::Event) {...}
}
```

I too don't have a strong opinion on this, for simplicity I would prefer to have it one block, but wouldn't mind have it split too, just that at the moment I don't see reason for it.

Reply:
Difference is if `ConnectionMonitorBehaviour` would work with anything that is not a `ConnectionMonitor`. This way we enforce the behaviour not to work under any other bound. But its true that for testing purposes relaxing that would be better. We cannot test the other methods separetely unless we have a mock `ConnectionMonitor`. But since that is tested already it should be fine. Whatever you feel in this case @bacv. But good point @ntn-x2 .